src/libopensc/card-gids.c

```diff
@@ -820,6 +849,54 @@ static int gids_set_security_env(sc_card_t *card,
 	return r;
 }
 
+
+static int
+gids_decipher(struct sc_card *card,
+		const u8 * crgram, size_t crgram_len,
+		u8 * out, size_t outlen)
+{
+	int r;
+	struct sc_apdu apdu;
+	u8 *sbuf = NULL;
+
+	if (card == NULL || crgram == NULL || out == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	LOG_FUNC_CALLED(card->ctx);
+	sc_log(card->ctx,
+	       "Gids decipher: in-len %"SC_FORMAT_LEN_SIZE_T"u, out-len %"SC_FORMAT_LEN_SIZE_T"u",
+	       crgram_len, outlen);
+
+	sbuf = malloc(crgram_len + 1);
+	if (sbuf == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	 * P1:  0x80  Resp: Plain value
+	 * P2:  0x86  Cmd: Padding indicator byte followed by cryptogram */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_4, 0x2A, 0x80, 0x86);
+	apdu.resp    = out;
+	apdu.resplen = outlen;
+	apdu.le      = outlen;
+
+	sbuf[0] = 0; /* padding indicator byte, 0x00 = No further indication */
+	memcpy(sbuf + 1, crgram, crgram_len);
+	apdu.data = sbuf + 1; /* Skip padding indication not needed unless SM */
```

Comment:
Do I see right that you add the padding indcator byte here and than not use it in the APDU?

Reply:
Yes, The GIDS docx talks about the PI being present, but the Microsoft "Virtual Smart Card  (and @vletoux  applet) will not work if the PI is present in decipher.  iso7816-4 hints that it may be optional. The GIDS docx says PI is needed if Secure Messaging is being used but the driver does not currently support SM. 

The gids_decipher was copied from iso7816.c which always adds a PI of 0x00. 
I left the code in case we ever add SM to the driver. debug logs show `certutil` does not use SM. and only works if the PI is not present.

I could just have easily not allocated the sbuf and used something like: `apdu.data = crgram;`

Suggestions?

Reply:
I do not know anything about GIDS, but if it is just more code (copied from ISO), which is not used, I would rather see it simplified without the needless allocations and copying. If the SM will be implemented some time, adding the padding indicator will probably be least work.